### PR TITLE
feat(egghead): spike search

### PR DESCRIPTION
**Pull request by vojtaholik**
- based on [this example](https://github.com/typesense/showcase-nextjs-instantsearch-next-app-router-ssr-steam-games-search) around SSR search using Typesense, Next.js App Router and Instantsearch
- shadcn inputs with instantsearch 7 hooks
- nuqs for search params state management
- routing is simplified and notion of topic pages is TBD

https://github.com/user-attachments/assets/030098fd-252a-4427-9ec5-f838eca1bcad

preview: https://course-builder-egghead-5fxxeijmx-eggheadio1.vercel.app/q?instructor=Joel+Hooks

gif

### CODE REVIEW

File: apps/egghead/src/app/(search)/q/_components/search.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	const [type, setType] = useQueryState('type')
+	const [instructor, setInstructor] = useQueryState('instructor')
+	const [query, setQuery] = useQueryState('q')
+
+	return (
+		<InstantSearchNext
+			searchClient={typesenseInstantsearchAdapter.searchClient}
+			indexName={TYPESENSE_COLLECTION_NAME}
+			routing={false}
+			onStateChange={({ uiState, setUiState }) => {
```
Comment (vojtaholik) [inline apps/egghead/src/app/(search)/q/_components/search.tsx:38]: all the routing can be done here. right now we're only working with search params.

File: apps/egghead/src/app/(search)/q/_components/search.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+				const query = uiState[TYPESENSE_COLLECTION_NAME]?.query
+				setQuery(query || null)
+
+				handleRefinementListChange('type', setType)
+				handleRefinementListChange('instructor_name', setInstructor)
+
+				// IMPORTANT: always call setUiState to sync the state
+				setUiState(uiState)
+			}}
```
Comment (vojtaholik) [inline apps/egghead/src/app/(search)/q/_components/search.tsx:62]: all the routing can be done here. right now we're only working with search params.

File: apps/egghead/src/app/(search)/q/page.tsx [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import Search from './_components/search'
+
+export const dynamic = 'force-dynamic'
```
Comment (vojtaholik) [inline apps/egghead/src/app/(search)/q/page.tsx:3]: ⚡

File: apps/egghead/src/app/(search)/q/_components/search.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+			future={{ preserveSharedStateOnUnmount: true }}
+		>
+			<SearchBox />
+			<div className="mb-3 flex flex-col flex-wrap items-center gap-3 sm:flex-row">
+				<RefinementList
+					attribute="instructor_name"
+					queryKey="instructor"
+					label="Instructor"
+				/>
+				<RefinementList attribute="type" label="Type" />
```
Comment (vojtaholik) [inline apps/egghead/src/app/(search)/q/_components/search.tsx:87]: these are all talking with typesense via instantsearch hooks

**Review by joelhooks — APPROVED** [review]
would be rad to see a version of this in the authoring environment

**Pull request merged**